o5logon dissector overhaul.

magnumripper opened the pull request:
This is needed for newer versions of Oracle. It also fixes various problems in the original code but ultimately I need a fix for #741 to do it good.

kholia commented:
Here are some .pcap files for testing these changes,

https://raw.githubusercontent.com/kholia/ctf/master/ggyy-99b3ed3ed5e2a293c697784ac94ee8c6.pcap

http://openwall.info/wiki/_media/john/Oracle-O5LOGON.tar.gz

And pcap files from https://github.com/hashcat/oclHashcat/issues/371

kholia commented:
@koeppea @LocutusOfBorg Ping, it would be great if you could review, and possibly merge this PR.

LocutusOfBorg commented:
I would like to see a fix for #741 before merging this one... isn''t it better?

koeppea commented:
I'd also like to add some annotations for this one. Could yet manage to afford the time.

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -1,43 +1,62 @@
+/* PKCS#7 padding used? */
+#define MAYBE 0
+#define YES   1
+#define NO    2
```
pre-processor directives and variables should be unique. 
please add somwhat like EC_O5LOGON_ to the pre-processor variables.

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -46,6 +65,24 @@ void o5logon_init(void);
 /************************************************/
 
+#undef memrchr
+#define memrchr my_memrchr
```
What is the purpose of this renaming? You shouldn't replace standard function names.
Why do you implement your own version of this standard function?

magnumripper replied:
That function is not present in eg. OSX so I implemented it. The define saves us on eg. Linux where it **is** present.

I don't understand cmake but I presume this could be dealt with in some more clever way - but this works fine as-is on Linux and OSX.

koeppea replied:
So we'd should only use it when being on a BSD or DARWIN a.k.a. MacOS operating system. The pre-processor variables are set by cmake.
I'd even recommend putting it into a "common" file like ec_utils.c that make it easier to be reused by other functions in future.
Or we could put it into a dedicated source / header file and include it into compilation processes depending on Cmake tests.

I could also help you with the Cmake part if you want.

koeppea replied:
Just checked FreeBSD. There the `memrchr()` function exists. MacOSX apparently is the only OS lacking `memrchr()` .

magnumripper replied:
Added 40979f2 which moves it to ec_utils.c and wraps with #if OS_DARWIN

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -62,106 +99,261 @@ FUNC_DECODER(dissector_o5logon)
+         return NULL;
+      }
+
+      u_char *ano;
```
Please declare variables only at the beginning of a block, otherwise this will generate a warning.

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -1,43 +1,62 @@
 #include <ec_dissect.h>
 #include <ec_session.h>
 
+//#define O5DEBUG
```
I'd rather recommend defining this (prefixed with EC_) with the value 0 for off and 1 for on.
to include such a block you the do `#if EC_O5DEBUG` instead of `#ifdef ...`.
But this is really somewhat cosmetic and flavor.

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -1,43 +1,62 @@
 struct o5logon_status {
-   u_char status;
    u_char user[129];
+   u_char srv_addr[22];
```
What is the purpose of `srv_addr[22]`? Holding the string representation of a IP address?
 1. it won't fit an IPv6 address
 2. managing IP addresses string or human representrationwise is somewhat painful.

I rather recommend using the type `struct ip_addr *` instead and use the **ip_addr_*** functions to initialize, handle and represent the IP address.

magnumripper commented:
Thank you for reviewing. Requested changes are now implemented and it's also rebased onto current tree. Tested on OSX (prior to this commit it was also tested on Linux).

magnumripper commented:
> I would like to see a fix for #741 before merging this one... isn''t it better?

It would be beneficial for JtR users to have it merged right away. If/when #741 is fixed I will promptly make a PR with enhancements to this dissector.

koeppea commented:
@magnumripper may I request contribution / write permissions to your fork?
I'd like to add a generic Cmake test-based approach to determine whether or not this function exist.
I can also move it to ec_strings.c as @gvanem suggested, as `memrchr()` is natively also defined in _string.h_.

magnumripper commented:
@koeppea you are now a collaborator.

koeppea commented:
Thanks.

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -62,106 +82,318 @@ FUNC_DECODER(dissector_o5logon)
+#if EC_O5LOGON_DEBUG
+         DISSECT_MSG("O5LOGON: client ver %u.%u.%u.%u.%u (PKCS)\n", 
+               ano[0], ano[1] >> 4, ano[1] & 15 , ano[2], ano[3]);
+#endif
```
Wouldn't this being worth to be print w/o DEBUG?

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -62,106 +82,318 @@ FUNC_DECODER(dissector_o5logon)
+                     ntohs(PACKET->L4.dst), 
+                     conn_status->user, 
+                     conn_status->flags.c_ano == 1 ? "" : "(no ANO seen)");
+#endif
```
This as well?

koeppea commented on src/dissectors/ec_o5logon.c:
```diff
@@ -62,106 +82,318 @@ FUNC_DECODER(dissector_o5logon)
+#if EC_O5LOGON_DEBUG
+                  DISSECT_MSG("O5LOGON: server ver %u.%u.%u.%u.%u\n", ano[0], 
+                        ano[1] >> 4, ano[1] & 15 , ano[2], ano[3]);
+#endif
```
This as well?

magnumripper replied:
I personally had no need for either of them except while debugging / trying to understand the protocol), but I'm fine with including them w/o debug.

koeppea replied:
What do you think about the amount of information in the messages pane:
```
$ ettercap -Tqr ~/Downloads/ggyy-99b3ed3ed5e2a293c697784ac94ee8c6.pcap 

ettercap 0.8.2 copyright 2001-2015 Ettercap Development Team

Reading from /home/koeppea/Downloads/ggyy-99b3ed3ed5e2a293c697784ac94ee8c6.pcap
Libnet failed IPv4 initialization. Don't send IPv4 packets.
Libnet failed IPv6 initialization. Don't send IPv6 packets.
  34 plugins
  42 protocol dissectors
  57 ports monitored
20530 mac vendor fingerprint
1766 tcp OS fingerprint
2182 known services

Starting Unified sniffing...

O5LOGON: AASH@192.168.131.1:$o5logon$209C317916169AEFD4ACA354CBA5588902933CBB2C31909898F1B4A7B1B4612A423FE59958FB8E07400FC69577B7BE34*ED3BE8DDEBF2FE6FB36F
O5LOGON: AASH@192.168.131.1:$o5logon$209C317916169AEFD4ACA354CBA5588902933CBB2C31909898F1B4A7B1B4612A423FE59958FB8E07400FC69577B7BE34*ED3BE8DDEBF2FE6FB36F
O5LOGON: AASH@192.168.131.1:$o5logon$209C317916169AEFD4ACA354CBA5588902933CBB2C31909898F1B4A7B1B4612A423FE59958FB8E07400FC69577B7BE34*ED3BE8DDEBF2FE6FB36F
O5LOGON: AASH@192.168.131.1:$o5logon$209C317916169AEFD4ACA354CBA5588902933CBB2C31909898F1B4A7B1B4612A423FE59958FB8E07400FC69577B7BE34*ED3BE8DDEBF2FE6FB36F
O5LOGON: AASH@192.168.131.1:$o5logon$209C317916169AEFD4ACA354CBA5588902933CBB2C31909898F1B4A7B1B4612A423FE59958FB8E07400FC69577B7BE34*ED3BE8DDEBF2FE6FB36F
....
```
What can be done with this information?

magnumripper replied:
Like I wrote below it's password hashes. I could try to reject dupes tho, but I'm not sure it's worth the effort. It shouldn't be a problem IRL.

koeppea replied:
@LocutusOfBorg what do you think about the deduplication? I have no feeling about the real world occurrence. 
On the other hand, we don't have such mechanisms on other dissectors and omitting it keeps the processing lightweight.

LocutusOfBorg replied:
I don't think it is worth the effort...

LocutusOfBorg replied:
somebody might want to sort and count how many people have the same password hash maybe?

koeppea commented:
Tested on Linux, MacOSX and FreeBSD.
@LocutusOfBorg: Good to go for me.

magnumripper commented:
> I'm not a Oracle expert, but I'd expected some less "verbose" information.

The output is password hashes, ready to crack in JtR or Hashcat.

magnumripper commented:
Oh, and thanks a lot for reviewing and fixing the last issues.

LocutusOfBorg commented on src/dissectors/ec_o5logon.c:
```diff
@@ -1,43 +1,63 @@
 #include <ec_session.h>
+#include <ec_utils.h>
+
+#define EC_O5LOGON_DEBUG       0
```
@koeppea what about "if not defined then define it to 0?" this way people might be able to feed this value from the environment, without having to change the source code.
it looks good to me, even if I didn't test it. Feel free to merge whenever you want :)

LocutusOfBorg commented:
merged thanks you all :)

regel commented:
@magnumripper This is awesome work. Thanks! To document the observed traffic, is it possible to summarize e.g. in a table the byte format observed on the wire versus the Oracle client and Oracle server versions actually in use?

magnumripper commented:
I wouldn't mind that, but unfortunately I don't have any time to invest.
